File: COA software/backend/app/schemas/simulation.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class ScenarioItemResponse(BaseModel):
    scenario_id: str
    name: str
    description: str
    difficulty: str
    train_count: int
    block_count: int


class SimulationRunRequest(BaseModel):
    scenario_id: Optional[str] = Field(default="SHARED_BLOCK_OPTIMIZATION", description="Scenario ID to load")
    plan_mode: Optional[str] = Field(default="AI_OPTIMIZED", description="MANUAL_BASELINE or AI_OPTIMIZED")


class SimulationControlRequest(BaseModel):
    delta_minutes: Optional[int] = Field(default=5, description="Minutes to step forward")
    speed_multiplier: Optional[float] = Field(default=1.0, description="1.0, 2.0, 5.0, 10.0")


class StationNodeSchema(BaseModel):
    station_id: str
    code: str
    name: str
    km_post: float
    tracks_count: int
    x_coord: float
    y_coord: float


class TrackSegmentSchema(BaseModel):
    segment_id: str
    from_station_id: str
    to_station_id: str
    length_km: float
    max_speed_kmh: int
    electrified: bool
    is_blocked: bool
    active_block_id: Optional[str] = None


class SimulatedTrainSchema(BaseModel):
    train_id: str
    train_number: str
    train_type: str
    direction: str
    origin: str
    destination: str
    current_station: Optional[str] = None
    current_section: Optional[str] = None
    progress_pct: float
    speed_kmh: float
    scheduled_departure: str
    scheduled_arrival: str
    actual_departure: Optional[str] = None
    actual_arrival: Optional[str] = None
    delay_minutes: float
    status: str


class SimulatedTaskSchema(BaseModel):
    task_id: str
    task_code: str
    department: str
    asset_id: str
    asset_name: str
    duration_minutes: int
    progress_pct: float
    status: str


class SimulatedBlockSchema(BaseModel):
    block_id: str
    corridor_id: str
    section_code: str
    start_time: str
    end_time: str
    duration_minutes: int
    departments: List[str]
    is_shared: bool
    status: str
    tasks: List[SimulatedTaskSchema] = []


class SimulationEventSchema(BaseModel):
    event_id: str
    event_type: str
    simulation_time: str
    title: str
    description: str
    severity: str


class SimulationMetricsSchema(BaseModel):
    active_trains: int
    completed_trains: int
    delayed_trains: int
    total_train_delay_minutes: float
    active_blocks: int
    completed_blocks: int
    active_maintenance_tasks: int
    completed_maintenance_tasks: int
    conflicts_detected: int
    conflicts_resolved: int
    asset_availability_pct: float
    block_utilization_pct: float


class SimulationStateResponse(BaseModel):
    simulation_id: str
    scenario_id: str
    scenario_name: str
    simulation_date: str
    simulation_time_minutes: int
    simulation_time_str: str
    status: str
    speed_multiplier: float
    plan_mode: str
    network_stations: List[StationNodeSchema] = []
    track_segments: List[TrackSegmentSchema] = []
    trains: List[SimulatedTrainSchema] = []
    blocks: List[SimulatedBlockSchema] = []
    events: List[SimulationEventSchema] = []
    metrics: SimulationMetricsSchema
    plan_comparison: Dict[str, Any] = {}

File: COA software/frontend/src/types/simulation.ts
export interface ScenarioItem {
  scenario_id: string
  name: string
  description: string
  difficulty: string
  train_count: number
  block_count: number
}

export interface StationNode {
  station_id: string
  code: string
  name: string
  km_post: number
  tracks_count: number
  x_coord: number
  y_coord: number
}

export interface TrackSegment {
  segment_id: string
  from_station_id: string
  to_station_id: string
  length_km: number
  max_speed_kmh: number
  electrified: boolean
  is_blocked: boolean
  active_block_id?: string
}

export interface SimulatedTrain {
  train_id: string
  train_number: string
  train_type: string
  direction: string
  origin: string
  destination: string
  current_station?: string
  current_section?: string
  progress_pct: number
  speed_kmh: number
  scheduled_departure: string
  scheduled_arrival: string
  actual_departure?: string
  actual_arrival?: string
  delay_minutes: number
  status: string
}

export interface SimulatedTask {
  task_id: string
  task_code: string
  department: string
  asset_id: string
  asset_name: string
  duration_minutes: number
  progress_pct: number
  status: string
}

export interface SimulatedBlock {
  block_id: string
  corridor_id: string
  section_code: string
  start_time: string
  end_time: string
  duration_minutes: number
  departments: string[]
  is_shared: boolean
  status: string
  tasks: SimulatedTask[]
}

export interface SimulationEvent {
  event_id: string
  event_type: string
  simulation_time: string
  title: string
  description: string
  severity: string
}

export interface SimulationMetrics {
  active_trains: number
  completed_trains: number
  delayed_trains: number
  total_train_delay_minutes: number
  active_blocks: number
  completed_blocks: number
  active_maintenance_tasks: number
  completed_maintenance_tasks: number
  conflicts_detected: number
  conflicts_resolved: number
  asset_availability_pct: number
  block_utilization_pct: number
}

export interface SimulationState {
  simulation_id: string
  scenario_id: string
  scenario_name: string
  simulation_date: string
  simulation_time_minutes: number
  simulation_time_str: string
  status: 'INITIALIZING' | 'RUNNING' | 'PAUSED' | 'COMPLETED' | 'RESET'
  speed_multiplier: number
  plan_mode: 'MANUAL_BASELINE' | 'AI_OPTIMIZED'
  network_stations: StationNode[]
  track_segments: TrackSegment[]
  trains: SimulatedTrain[]
  blocks: SimulatedBlock[]
  events: SimulationEvent[]
  metrics: SimulationMetrics
  plan_comparison: any
}
